libkbfs/tlf_edit_history.go

```diff
@@ -0,0 +1,355 @@
+	Filepath  string // relative to the TLF root
+	Type      TlfEditNotificationType
+	LocalTime time.Time // reflects difference between server and local clock
+}
+
+const (
+	// How many edits per writer we want to return in the complete history?
+	desiredEditsPerWriter = 20
+
+	// How far back we're willing to go to get the complete history.
+	maxMDsToInspect = 1000
+)
+
+// TlfEditList is a list of edits by a particular user, that can be
+// sort by increasing timestamp.
+type TlfEditList []TlfEdit
+
+// Len implements sort.Interface for TlfEditList
+func (tel TlfEditList) Len() int {
+	return len(tel)
+}
+
+// Less implements sort.Interface for TlfEditList
+func (tel TlfEditList) Less(i, j int) bool {
+	return tel[i].LocalTime.Before(tel[j].LocalTime)
+}
+
+// Swap implements sort.Interface for TlfEditList
+func (tel TlfEditList) Swap(i, j int) {
+	tel[j], tel[i] = tel[i], tel[j]
+}
+
+// TlfWriterEdits is a map of a writer name to the most recent file
+// edits in a given folder by that writer.
+type TlfWriterEdits map[keybase1.UID]TlfEditList
+
+func (we TlfWriterEdits) isComplete() bool {
+	for _, edits := range we {
+		if len(edits) < desiredEditsPerWriter {
+			return false
+		}
+	}
+	return true
+}
+
+type writerEditEstimates map[keybase1.UID]int
+
+func (wee writerEditEstimates) isComplete() bool {
+	for _, count := range wee {
+		if count < desiredEditsPerWriter {
+			return false
+		}
+	}
+	return true
+}
+
+func (wee *writerEditEstimates) update(rmds []ImmutableRootMetadata) {
+	for i := len(rmds) - 1; i >= 0; i-- {
+		rmd := rmds[i]
+		if rmd.IsWriterMetadataCopiedSet() {
+			continue
+		}
+		writer := rmd.LastModifyingWriter
+		for _, op := range rmd.data.Changes.Ops {
+			// Estimate the number of writes just based on operations
+			// (without yet taking into account whether the same file
+			// is being edited more than once).
+			switch realOp := op.(type) {
+			case *createOp:
+				if realOp.Type == Dir || realOp.Type == Sym {
+					continue
+				}
+				(*wee)[writer]++
+			case *syncOp:
+				(*wee)[writer]++
+			}
+		}
+	}
+}
+
+func (wee *writerEditEstimates) reset(edits TlfWriterEdits) {
+	for writer := range *wee {
+		(*wee)[writer] = len(edits[writer])
+	}
+}
+
+// TlfEditHistory allows you to get the update history about a
+// particular TLF.
+type TlfEditHistory struct {
+	config Config
+	fbo    *folderBranchOps
+	log    logger.Logger
+
+	lock  sync.Mutex
+	edits TlfWriterEdits
+}
+
+func (teh *TlfEditHistory) getEditsCopyLocked() TlfWriterEdits {
+	if teh.edits == nil {
+		return nil
+	}
+	edits := make(TlfWriterEdits)
+	for user, userEdits := range teh.edits {
+		userEditsCopy := make([]TlfEdit, len(userEdits))
+		copy(userEditsCopy, userEdits)
+		edits[user] = userEditsCopy
+	}
+	return edits
+}
+
+func (teh *TlfEditHistory) getEditsCopy() TlfWriterEdits {
+	teh.lock.Lock()
+	defer teh.lock.Unlock()
+	return teh.getEditsCopyLocked()
+}
+
+func (teh *TlfEditHistory) updateRmds(rmds []ImmutableRootMetadata,
```

Comment:
`rmds []Immutable, olderRmds ...` -> `rmds, olderRmds []Immutable...`

Also, any particular reason you're not using append? Is it to avoid hidden sharing? If so, probably worth a comment.

Also, may as well get rid of the error param, since it's always nil.